common/src/main/java/com/microsoft/identity/common/crypto/AndroidWrappedKeyLoader.java

```diff
@@ -0,0 +1,335 @@
+
+        Logger.info(TAG + methodName, "New key is generated with thumbprint: " +
+                KeyUtil.getKeyThumbPrint(key));
+
+        return key;
+    }
+
+    /**
+     * Load the saved keystore-encrypted key. Will only do read operation.
+     *
+     * @return SecretKey. Null if there isn't any.
+     */
+    @Nullable
+    /* package */ SecretKey readSecretKeyFromStorage() throws ClientException {
+        final String methodName = ":readSecretKeyFromStorage";
+        final KeyPair keyPair;
+        try {
+            keyPair = readKeyStoreKeyPair();
+            if (keyPair == null) {
+                Logger.warn(TAG + methodName, "key does not exist in keystore");
+                deleteSecretKeyFromStorage();
+                return null;
+            }
+
+            final byte[] wrappedSecretKey = FileUtil.readFromFile(mContext, KEY_FILE_PATH, KEY_FILE_SIZE);
+            if (wrappedSecretKey == null) {
+                Logger.warn(TAG + methodName, "Key file is empty");
+                deleteSecretKeyFromStorage();
+                return null;
+            }
+
+            final SecretKey key = AndroidKeyStoreUtil.unwrap(wrappedSecretKey, getKeySpecAlgorithm(), keyPair, WRAP_ALGORITHM);
+
+            Logger.info(TAG + methodName, "New key is generated with thumbprint: " +
+                    KeyUtil.getKeyThumbPrint(key));
+
+            return key;
+        } catch (ClientException e) {
+            // Reset KeyPair info so that new request will generate correct KeyPairs.
+            // All tokens with previous SecretKey are not possible to decrypt.
+            Logger.warn(TAG + methodName, "Error when loading key from Keystore, " +
+                    "wipe all existing key data ");
+            deleteSecretKeyFromStorage();
+            throw e;
+        }
+    }
+
+    /**
+     * Encrypt the given unencrypted symmetric key with Keystore key and save to storage.
+     */
+    private void saveSecretKeyToStorage(@NonNull SecretKey unencryptedKey) throws ClientException {
+        final KeyPair keyPair = generateKeyStoreKeyPair();
+        final byte[] keyWrapped = AndroidKeyStoreUtil.wrap(unencryptedKey, keyPair, WRAP_ALGORITHM);
+        FileUtil.writeDataToFile(keyWrapped, mContext, KEY_FILE_PATH);
+    }
+
+    /**
+     * Wipe all the data associated from this key.
+     */
+    private void deleteSecretKeyFromStorage() throws ClientException {
+        AndroidKeyStoreUtil.deleteKey(mAlias);
+        FileUtil.deleteFile(mContext, KEY_FILE_PATH);
+        mCachedKey = null;
+    }
+
+    /**
+     * Generate the key in {@link KeyStore}.
+     *
+     * @return KeyPair. Null if there isn't any.
+     */
+    @NonNull
+    private synchronized KeyPair generateKeyStoreKeyPair()
+            throws ClientException {
+        final String methodName = ":generateKeyStoreKeyPair";
+        try {
+            logFlowStart(methodName, AuthenticationConstants.TelemetryEvents.KEYCHAIN_WRITE_START);
```

Comment:
A bit confused about these telemetry names... we're not actually communicating with `KeyChain` here are we?